dialed_addrs.go

```diff
@@ -0,0 +1,151 @@
+package autonat
+
+import (
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peerstore"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// ActivationThresh sets how many times an address must be seen as "activated"
+// and therefore marked as confirmed.
+// The "seen" events expire by default after 40 minutes
+// (OwnObservedAddressTTL * ActivationThreshold). They are cleaned up during
+// GC.
+var activationThresh = 4
+
+// observation records an address observation from an "observer" (where every IP
+// address is a unique observer).
+type observation struct {
+	// seenTime is the last time this observation was made.
+	seenTime time.Time
+}
+
+// observedAddr is an entry for an address reported by AutoNAT servers.
+// We only use addresses that:
+// - have been observed at least 4 times in last 40 minutes.
+// - have been observed at least once recently (10 minutes), because our position in the
+//   network, or network port mapppings, may have changed.
+type observedAddr struct {
+	addr       ma.Multiaddr
+	seenBy     map[string]observation // peer(observer) address -> observation info
+	lastSeen   time.Time
+	numInbound int
+}
+
+func (oa *observedAddr) activated() bool {
+
+	// We only activate if other peers observed the same address
+	// of ours at least 4 times. SeenBy peers are removed by GC if
```

Comment:
this seems high? can we activate after 1 with no false negatives, but have a confidence tracking how sure we are?